docs/design/gh-ost-integration.md

```diff
@@ -0,0 +1,509 @@
+
+And a new task payload type and task executor for each execution task.
+
+```Go
+// issue related
+
+const IssueDatabaseSchemaUpdateGhost IssueType = "bb.issue.database.schema.update.ghost"
+
+type UpdateSchemaGhostDetail struct {
+    DatabaseID int `json:"databaseId"`
+    DatabaseName string `json:"databaseName"`
+    Statement string `json:"statement"`
+    EarliestAllowedTs int64 `jsonapi:"attr,earliestAllowedTs"`
+}
+
+type UpdateSchemaGhostContext struct {
+    MigrationType db.MigrationType `json:"migrationType"`
+    UpdateSchemaDetailList []*UpdateSchemaGhostDetail `json:"updateSchemaDetailList`
+    VCSPushEvent *vcs.PushEvent
+}
+
+// task related
+
+const TaskDatabaseSchemaUpdateGhost TaskType = "bb.task.database.schema.update.ghost"
+type TaskDatabaseSchemaUpdateGhostPayload struct {
+    MigrationType db.MigrationType `json:"migrationType,omitempty"`
+    Statement     string           `json:"statement,omitempty"`
+    SchemaVersion string           `json:"schemaVersion,omitempty"`
+    VCSPushEvent  *vcs.PushEvent   `json:"pushEvent,omitempty"`
+    // more to come
+}
+
+const TaskDatabaseSchemaUpdateGhostCutover = "bb.task.database.schema.update.ghost.cutover"
+type TaskDatabaseSchemaUpdateGhostCutoverPayload struct {
+    // ParentTaskID references back to the gh-ost executor task ID
+    ParentTaskID int `json:"parentTaskId,omitempty"`
+}
+
+const TaskDatabaseSchemaUpdateGhostDropOriginalTable = "bb.task.database.schema.update.ghost.drop-original-table"
+type TaskDatabaseSchemaUpdateGhostDropOriginalTablePayload struct {
+    DatabaseName string `json:"databaseName,omitempty"`
+    TableName string `json:"tableName,omitempty"`
+}
+```
+
+### gh-ost executor
+
+#### How migration works now
+
+Here's how bytebase executes migration now.
+
+1. `server/task_scheduler.go` schedules a migration task.
+2. `server/task_executor_schema_update.go` `func RunOnce()`
+3. `server/task_executor.go` `func runMigration()`
+4. `driver.ExecuteMigration()`
+5. `util.ExecuteMigration()`
+
+`util.ExecuteMigration` basically does the following,
+
+```Go
+func ExecuteMigration(ctx context.Context, m *db.MigrationInfo, statement string) (int64, string, error) {
+    preMigration()
+    defer postMigration()
+    Execute(statement)
+}
+```
+
+#### Use gh-ost to execute migration
+
+Since `Execute(statement)` merely just executes the DDL statement, it should be interchangeable with gh-ost executing the migration.
+
+We will refactor `ExecuteMigration` and export `beginMigration` and `endMigration`, so with gh-ost it will be `beginMigration`->`Migrate`->`endMigration`.
+
+```Go
+// task_executor_schema_update_ghost.go
+
+type SchemaUpdateTaskExecutorGhost struct {
+    l *zap.Logger
+}
+
+func (exec *SchemaUpdateTaskExecutor) RunOnce(ctx context.Context, server *Server, task *api.Task, channel <-chan string) (terminated bool, result *api.TaskRunResultPayload, err error) {
+    PreMigration()
```

Comment:
Probably not as part of single RunOnce, PreMigration+ghostExecuteMigration will be in the first task, PostMigration will be part of the last task after deleting the del table.